server/schedulers/adjacent_region.go

```diff
@@ -0,0 +1,255 @@
+		}
+		return newBalanceAdjacentRegionScheduler(opt, limiter), nil
+	})
+}
+
+type balanceAdjacentRegionScheduler struct {
+	*baseScheduler
+	opt         schedule.Options
+	selector    schedule.Selector
+	leaderLimit uint64
+	peerLimit   uint64
+	lastKey     []byte
+	ids         []uint64
+}
+
+// newBalanceAdjacentRegionScheduler creates a scheduler that tends to disperse adjacent region
+// on each store.
+func newBalanceAdjacentRegionScheduler(opt schedule.Options, limiter *schedule.Limiter, args ...uint64) schedule.Scheduler {
+	filters := []schedule.Filter{
+		schedule.NewBlockFilter(),
+		schedule.NewStateFilter(opt),
+		schedule.NewHealthFilter(opt),
+		schedule.NewSnapshotCountFilter(opt),
+		schedule.NewStorageThresholdFilter(opt),
+	}
+	base := newBaseScheduler(limiter)
+	s := &balanceAdjacentRegionScheduler{
+		baseScheduler: base,
+		opt:           opt,
+		selector:      schedule.NewRandomSelector(filters),
+		leaderLimit:   defaultAdjacentLeaderLimit,
+		peerLimit:     defaultAdjacentPeerLimit,
+		lastKey:       []byte(""),
+	}
+	if len(args) == 2 {
+		s.leaderLimit = args[0]
+		s.peerLimit = args[1]
+	}
+	return s
+}
+
+func (l *balanceAdjacentRegionScheduler) GetName() string {
+	return "balance-adjacent-region-scheduler"
+}
+
+func (l *balanceAdjacentRegionScheduler) GetType() string {
+	return "adjacent-region"
+}
+
+func (l *balanceAdjacentRegionScheduler) GetMinInterval() time.Duration {
+	return minAdjacentSchedulerInterval
+}
+
+func (l *balanceAdjacentRegionScheduler) GetNextInterval(interval time.Duration) time.Duration {
+	return intervalGrow(interval, maxAdjacentSchedulerInterval, linearGrowth)
+}
+
+func (l *balanceAdjacentRegionScheduler) IsScheduleAllowed() bool {
+	return l.allowBalanceLeader() || l.allowBalanceLeader()
+}
+
+func (l *balanceAdjacentRegionScheduler) allowBalanceLeader() bool {
+	return l.limiter.OperatorCount(core.AdjacentLeaderKind) < l.leaderLimit
+}
+
+func (l *balanceAdjacentRegionScheduler) allowBalancePeer() bool {
+	return l.limiter.OperatorCount(core.AdjacentPeerKind) < l.peerLimit
+}
+
+func (l *balanceAdjacentRegionScheduler) Schedule(cluster schedule.Cluster) *schedule.Operator {
+	if l.ids == nil {
+		l.ids = make([]uint64, 0, len(cluster.GetStores()))
+	}
+
+	regions := cluster.ScanRegions(l.lastKey, scanLimit)
+	adjacentRegions := make([]*core.RegionInfo, 0, scanLimit)
+	for _, r := range regions {
+		l.lastKey = r.StartKey
+		if len(adjacentRegions) == 0 {
+			adjacentRegions = append(adjacentRegions, r)
+			continue
+		}
+
+		// append if the region are adjacent
+		lastRegion := adjacentRegions[len(adjacentRegions)-1]
+		if lastRegion.Leader.GetStoreId() == r.Leader.GetStoreId() && bytes.Equal(lastRegion.EndKey, r.StartKey) {
+			adjacentRegions = append(adjacentRegions, r)
+			continue
+		}
+
+		if len(adjacentRegions) == 1 {
+			adjacentRegions[0] = r
+		} else {
+			// got an adjacent regions
+			break
```

Comment:
So the maximum length of `adjacentRegions` is 2? Why does it need capacity _scanLimit_?

Reply:
we will append all adjacent regions, so it isn't 2. now I change it to record the max length adjacent regions.